test, refactor: add GetTransaction() coverage, improve rpc_rawtransaction

jonatack opened the pull request:
Following up on https://github.com/bitcoin/bitcoin/pull/22383#pullrequestreview-698583510, this pull adds missing `src/node/transaction::GetTransaction()` test coverage for combinations of `-txindex` and `blockhash` and does some refactoring of the test file.

jonatack committed:
Test src/node/transaction::GetTransaction() without -txindex

jonatack committed:
test: run 2nd getrawtransaction section with/without -txindex

(and make the 'string "Flase"' test clearer as requested by reviewers)

jonatack committed:
refactor: dedup/reorg createrawtransaction sequence number tests

jonatack committed:
test: add and harmonize getrawtransaction logging

jonatack committed:
test: remove no longer needed (ASCII art) comments

jonatack committed:
move-only: regroup similar rpc_rawtransaction tests together

jonatack committed:
refactor: separate the rpc_rawtransaction tests into functions

jonatack committed:
refactor: txid to constant in rpc_rawtransaction to isolate tests

jonatack committed:
test, refactor: rpc_rawtransaction PEP8

jonatack commented:
Test output after these changes.  The slow legacy multisig tests are placed at the end.

```rake
$ test/functional/rpc_rawtransaction.py 
2021-07-13T13:24:28.484000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_c86n_kg4
2021-07-13T13:24:29.836000Z TestFramework (INFO): Prepare some coins for multiple *rawtransaction commands
2021-07-13T13:24:42.694000Z TestFramework (INFO): Test getrawtransaction with -txindex, without -blocksonly
2021-07-13T13:24:42.856000Z TestFramework (INFO): Test getrawtransaction without -txindex, without -blocksonly
2021-07-13T13:24:42.981000Z TestFramework (INFO): Test getrawtransaction without -txindex, with -blocksonly
2021-07-13T13:24:43.116000Z TestFramework (INFO): Test getrawtransaction on genesis block coinbase returns an error
2021-07-13T13:24:43.121000Z TestFramework (INFO): Test createrawtransaction
2021-07-13T13:24:43.230000Z TestFramework (INFO): Test sendrawtransaction with missing input
2021-07-13T13:24:43.302000Z TestFramework (INFO): Test sendrawtransaction with missing prevtx info (bech32)
2021-07-13T13:24:43.378000Z TestFramework (INFO): Test sendrawtransaction with missing prevtx info (p2sh-segwit)
2021-07-13T13:24:43.455000Z TestFramework (INFO): Test sendrawtransaction with missing prevtx info (legacy)
2021-07-13T13:24:43.507000Z TestFramework (INFO): Test sendrawtransaction/testmempoolaccept with maxfeerate
2021-07-13T13:24:57.111000Z TestFramework (INFO): Test decoderawtransaction
2021-07-13T13:24:57.123000Z TestFramework (INFO): Test transaction version numbers
2021-07-13T13:24:57.129000Z TestFramework (INFO): Test raw multisig transactions (legacy)
2021-07-13T13:25:17.516000Z TestFramework (INFO): Stopping nodes
2021-07-13T13:25:17.776000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_c86n_kg4 on exit
2021-07-13T13:25:17.776000Z TestFramework (INFO): Tests successful
```

jonatack commented:
Test output after these changes.  The slow legacy multisig tests are placed at the end.

```rake
$ test/functional/rpc_rawtransaction.py 
2021-07-13T13:42:55.022000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_u_56kycw
2021-07-13T13:42:56.672000Z TestFramework (INFO): Prepare some coins for multiple *rawtransaction commands
2021-07-13T13:43:04.132000Z TestFramework (INFO): Test getrawtransaction with -txindex, without -blocksonly
2021-07-13T13:43:04.305000Z TestFramework (INFO): Test getrawtransaction without -txindex, without -blocksonly
2021-07-13T13:43:04.447000Z TestFramework (INFO): Test getrawtransaction without -txindex, with -blocksonly
2021-07-13T13:43:04.603000Z TestFramework (INFO): Test getrawtransaction with -txindex, with -blocksonly
2021-07-13T13:43:04.763000Z TestFramework (INFO): Test getrawtransaction on genesis block coinbase returns an error
2021-07-13T13:43:04.770000Z TestFramework (INFO): Test createrawtransaction
2021-07-13T13:43:04.917000Z TestFramework (INFO): Test sendrawtransaction with missing input
2021-07-13T13:43:04.969000Z TestFramework (INFO): Test sendrawtransaction with missing prevtx info (bech32)
2021-07-13T13:43:05.055000Z TestFramework (INFO): Test sendrawtransaction with missing prevtx info (p2sh-segwit)
2021-07-13T13:43:05.121000Z TestFramework (INFO): Test sendrawtransaction with missing prevtx info (legacy)
2021-07-13T13:43:05.165000Z TestFramework (INFO): Test sendrawtransaction/testmempoolaccept with maxfeerate
2021-07-13T13:43:07.578000Z TestFramework (INFO): Test decoderawtransaction
2021-07-13T13:43:07.598000Z TestFramework (INFO): Test transaction version numbers
2021-07-13T13:43:07.602000Z TestFramework (INFO): Test raw multisig transactions (legacy)
2021-07-13T13:43:13.866000Z TestFramework (INFO): Stopping nodes
2021-07-13T13:43:14.133000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_u_56kycw on exit
2021-07-13T13:43:14.134000Z TestFramework (INFO): Tests successful
```

mjdietzx commented:
ACK 993189b1fe39cfc29e960ea3a20092309001fa8f very nicely done!

jonatack commented:
Thanks @mjdietzx!  Rebased to master following the merge of #22447, dropping the first two commits; no other change.

`git range-diff 97153a7 993189b d27edf1`

kiminuo commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -69,24 +74,129 @@ def setup_network(self):
         self.connect_nodes(0, 2)
 
+    def sync_peers(self):
+        self.sync_mempools(self.nodes[0:3])
```
Question: Is this intentionally only on 4 nodes instead of 6? If it is, maybe `sync_peers` should be called slightly differently not to give the wrong impression.

jonatack replied:
> Question: Is this intentionally only on 4 nodes instead of 6?

Yes, as only nodes 0 to 3 have a mempool. The last two nodes are -blocksonly and `sync_mempools()` won't work on them.  Added a comment to clarify this.

```diff
-        self.sync_mempools()
+        self.sync_mempools(self.nodes[0:3])  # nodes 0 to 3 have a mempool
```

kiminuo commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -69,24 +74,129 @@ def setup_network(self):
+        vout = find_vout_for_address(self.nodes[1], txid, addr)
+        rawTx = self.nodes[1].createrawtransaction([{'txid': txid, 'vout': vout}], {self.nodes[1].getnewaddress(): 9.999})
+        rawTxSigned = self.nodes[1].signrawtransactionwithwallet(rawTx)
+        txId = self.nodes[1].sendrawtransaction(rawTxSigned['hex'])
```
Not originally your code, but variables named `txid` and `txId` are very easy to confuse, I think.

jonatack replied:
I initially planned to respond that this was out of scope, then saw this very issue had tripped me up (https://github.com/bitcoin/bitcoin/pull/22437#discussion_r684653116). Agree! Appended a final commit to change them to `txid1`, `txid2` when `txId` was being reused/reassigned in the same test function. Doing this also revealed an unused `txId` assignment; fixed.

kiminuo commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -69,24 +74,129 @@ def setup_network(self):
+            else:
+                # without -txindex
+                for verbose in [None, 0, False, 1, True]:
+                    assert_raises_rpc_error(-5, err_msg, self.nodes[n].getrawtransaction, txid, verbose)
```
I would expect `txId` instead of `txid` here. Or can you explain this please?

jonatack replied:
Good catch! This illustrates why using `txid` and `txId` variable names in the same test isn't ideal. Fixed in commit "Test src/validation::GetTransaction() with -blocksonly".

lsilva01 reviewed (approved):
Tested ACK https://github.com/bitcoin/bitcoin/pull/22437/commits/d27edf1d85f44c58029e0197251cca56873b860f on Ubuntu 20.04

rajarshimaitra commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -69,24 +74,129 @@ def setup_network(self):
+        rawTx = self.nodes[1].createrawtransaction([{'txid': txid, 'vout': vout}], {self.nodes[1].getnewaddress(): 9.999})
+        rawTxSigned = self.nodes[1].signrawtransactionwithwallet(rawTx)
+        txId = self.nodes[1].sendrawtransaction(rawTxSigned['hex'])
+        self.generate_and_sync(node=0, blocks=1)
```
I am not sure whether having two `txid` and `txId` transactions helps here, given both of them are confirmed. Because of that we are also not getting a test for only `mempool` transactions.

So maybe we can have a `txid` confirmed in a block, and then have `txId` in the mempool to check for both the scenario?

Will make the purpose of two transactions clear.

Also, the names are confusing, something more distinct would be helpful.

jonatack replied:
Interesting.  We use and need all three transactions, AFAICT.  Updated the names in the last commit to txid, txid2, and txid3.  The test still passes if we sync_blocks() for the second and third transactions instead of sync_all(), but I'd rather not mess with this here.

jonatack replied:
Took a fresh look this morning, good call!  The first conditional in `CTransactionRef GetTransaction` was not being exercised by the tests.
```c++
    if (mempool && !block_index) {
        CTransactionRef ptx = mempool->get(hash);
        if (ptx) return ptx;
    }
```
Added this assert that covers it:
```python
             # 1. valid parameters - supply txid along with various valid values for verbose
+            if n == 0 or n == 3:
+                # test with a tx in mempool, with and without -txindex
+                tx_in_mempool = self.nodes[n].sendtoaddress(self.nodes[n + 1].getnewaddress(), 0.1)
+                self.nodes[n].getrawtransaction(tx_in_mempool, True)
```

jonatack replied:
(Verified by removing that code in GetTransaction() to ensure the new assert fails without it and passes with it.)

rajarshimaitra reviewed (commented):
tACK https://github.com/bitcoin/bitcoin/pull/22437/commits/d27edf1d85f44c58029e0197251cca56873b860f.

verified that the test passes with #22383 changes.

Below are few comments regarding test coverage and arrangements.

practicalswift commented:
Concept ACK

jonatack commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -69,24 +74,129 @@ def setup_network(self):
+                for verbose in [None, 0, False, 1, True]:
+                    assert_raises_rpc_error(-5, err_msg, self.nodes[n].getrawtransaction, txid, verbose)
+
+            # 2. invalid parameters - supply txid and string "Flase"
```
You are right that it's unclear. Changed to the following in the first commit touching the code, "Run getrawtransaction tests without -txindex"
```diff
-            # 6. invalid parameters - supply txid and string "Flase"
-            assert_raises_rpc_error(-1, "not a boolean", self.nodes[n].getrawtransaction, txId, "Flase")
+            # 6. invalid parameters - supply txid and invalid boolean values (strings) for verbose
+            for value in ["True", "False"]:
+                assert_raises_rpc_error(-1, "not a boolean", self.nodes[n].getrawtransaction, txid=txId, verbose=value)
```

jonatack commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -69,24 +74,129 @@ def setup_network(self):
+            # 4. invalid parameters - supply txid and empty dict
+            assert_raises_rpc_error(-1, "not a boolean", self.nodes[n].getrawtransaction, txId, {})
+
+            # 5. invalid parameters - sequence out of range
```
Good idea! Done in the 3rd commit "refactor: dedup/reorg createrawtransaction sequence number tests".

jonatack commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -69,24 +74,129 @@ def setup_network(self):
-
-        self.log.info('Test getrawtransaction on genesis block coinbase returns an error')
+        self.log.info("Prepare some coins for multiple *rawtransaction commands")
+        self.generate_and_sync(node=2, blocks=1)
```
Well spotted! Thanks to the transactions in the next lines, it's not needed and might be an example of the accumulated layers of change in this test file. Removed in commit "refactor: extract generate/sync functions in rpc_rawtransaction" with an explanation in the commit message.

jonatack commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -69,24 +74,129 @@ def setup_network(self):
+                    # this test every time the output format changes.
+                    gottx = self.nodes[n].getrawtransaction(txId, verbose)['hex']
+                    assert_equal(gottx, rawTxSigned['hex'])
+                    assert 'in_active_chain' not in gottx
```
This assert tests that the `in_active_chain` field is absent as expected when the blockhash argument is not passed--see `src/rpc/rawtransaction.cpp::getrawtransaction()` or`bitcoin-cli help getrawtransaction`.

kiminuo replied:
Just to make myself more clear: L140 ends with `['hex']` so presumably `gottx` is a string, so testing on L142 whether `gottx` contains `in_active_chain` key seems always true.

But maybe I just didn't get your reply. Anyway, if I got it wrong, feel free to just ignore my comment.

jonatack replied:
Thank you for re-explaining. You are right! Fixing. Thank you!

jonatack replied:
Updated commit 4e83843f03dad710a0fad7a706e2618a7a69a6b3 to fix. Thanks @kiminuo!

jonatack commented:
Rebased due to #22510 and updated with the review feedback from @kiminuo and @rajarshimaitra (thanks!) 

Commit-by-commit changes:

`git range-diff db94d74 d27edf1 afb4eab`

Thank you @mjdietzx, @lsilva01, and @rajarshimaitra for the ACKs. Would you mind re-ACKing?

rajarshimaitra reviewed (commented):
Thanks @jonatack for considering the comments. 

Re tested ACK https://github.com/bitcoin/bitcoin/pull/22437/commits/afb4eab7a3d309c34357aa6ac54b03bebd15adfb

It might not be a good place to discuss test approaches here, but I have the following observations
- I am finding that `getrawtransaction_tests` setup can be simplified a bit. Instead of 3 transactions (all in the chain) we can do the test with just one.  All we need is a tx in a block and its id. So if we simply start the test like this 
 ```python
        test_tx = self.nodes[2].sendtoaddress(self.nodes[1].getnewaddress(), 1)
        test_tx_hex = self.nodes[2].getrawtransaction(test_tx)
        block1, block2 = self.nodes[2].generate(2)
        self.sync_peers()
 ```
   this will give us everything we want to do the assertions. This I feel will simplify the testing logic and would make the test easy 
   to reason about.   

- The #22383 changes behaviour that if `txindex` is on, the `blockhash` searching won't take place even if it's provided. It seems to me that this particular behaviour is not being tested. I am also not sure what can be a possible approach to test this. But I feel this needs to be covered, as it's a performance boost, and we don't want future PR to accidentally change this.

Would like to have your thoughts on the above. It's not necessary to have those changes in this PR itself.  

Below is a minor redundancy I found.

jonatack commented:
@rajarshimaitra I agree, bringing together the various related tests shows that we can simplify them. I'll look at integrating the following diff based on your suggestion, which works for me, into the changes. 

<details><summary>code diff</summary><p>


```diff
diff --git a/test/functional/rpc_rawtransaction.py b/test/functional/rpc_rawtransaction.py
index 84210d3a03..cc6324d9fc 100755
--- a/test/functional/rpc_rawtransaction.py
+++ b/test/functional/rpc_rawtransaction.py
@@ -101,19 +101,9 @@ class RawTransactionsTest(BitcoinTestFramework):
             self.raw_multisig_transaction_legacy_tests()
 
     def getrawtransaction_tests(self):
-        addr = self.nodes[1].getnewaddress()
-        txid = self.nodes[0].sendtoaddress(addr, 10)
-        self.generate_and_sync(node=0, blocks=1)
-        vout = find_vout_for_address(self.nodes[1], txid, addr)
-        rawTx = self.nodes[1].createrawtransaction([{'txid': txid, 'vout': vout}], {self.nodes[1].getnewaddress(): 9.999})
-        rawTxSigned = self.nodes[1].signrawtransactionwithwallet(rawTx)
-        txid2 = self.nodes[1].sendrawtransaction(rawTxSigned['hex'])
-        self.generate_and_sync(node=0, blocks=1)
-
-        # Make a tx by sending, then generate 2 blocks; block1 has the tx in it
-        txid3 = self.nodes[2].sendtoaddress(self.nodes[1].getnewaddress(), 1)
+        tx = self.nodes[2].sendtoaddress(self.nodes[1].getnewaddress(), 1)
+        hex = self.nodes[2].getrawtransaction(tx)
         block1, block2 = self.nodes[2].generate(2)
-        self.sync_peers()
 
         err_msg = (
             "No such mempool transaction. Use -txindex or provide a block hash to enable"
@@ -134,47 +124,44 @@ class RawTransactionsTest(BitcoinTestFramework):
             if n == 0 or n == 5:
                 # with -txindex
                 for verbose in [None, 0, False]:
-                    assert_equal(self.nodes[n].getrawtransaction(txid2, verbose), rawTxSigned['hex'])
+                    assert_equal(self.nodes[n].getrawtransaction(tx, verbose), hex)
                 for verbose in [1, True]:
-                    gottx1 = self.nodes[n].getrawtransaction(txid2, verbose)
-                    assert_equal(gottx1['hex'], rawTxSigned['hex'])
-                    assert 'in_active_chain' not in gottx1.keys()
-                    gottx2 = self.nodes[n].getrawtransaction(txid=txid3, verbose=verbose)
-                    assert_equal(gottx2['txid'], txid3)
-                    assert 'in_active_chain' not in gottx2.keys()
+                    gottx = self.nodes[n].getrawtransaction(tx, verbose)
+                    assert_equal(gottx['hex'], hex)
+                    assert 'in_active_chain' not in gottx.keys()
             else:
                 # without -txindex
                 for verbose in [None, 0, False, 1, True]:
-                    assert_raises_rpc_error(-5, err_msg, self.nodes[n].getrawtransaction, txid2, verbose)
+                    assert_raises_rpc_error(-5, err_msg, self.nodes[n].getrawtransaction, tx, verbose)
 
             # 2. invalid parameters - supply txid and invalid boolean values (strings) for verbose
             for value in ["True", "False"]:
-                assert_raises_rpc_error(-1, "not a boolean", self.nodes[n].getrawtransaction, txid=txid2, verbose=value)
+                assert_raises_rpc_error(-1, "not a boolean", self.nodes[n].getrawtransaction, txid=tx, verbose=value)
 
             # 3. invalid parameters - supply txid and empty array
-            assert_raises_rpc_error(-1, "not a boolean", self.nodes[n].getrawtransaction, txid2, [])
+            assert_raises_rpc_error(-1, "not a boolean", self.nodes[n].getrawtransaction, tx, [])
 
             # 4. invalid parameters - supply txid and empty dict
-            assert_raises_rpc_error(-1, "not a boolean", self.nodes[n].getrawtransaction, txid2, {})
+            assert_raises_rpc_error(-1, "not a boolean", self.nodes[n].getrawtransaction, tx, {})
 
             # 5. with block hash
             # We should be able to get the raw transaction by providing the correct block
-            gottx = self.nodes[n].getrawtransaction(txid=txid3, verbose=True, blockhash=block1)
-            assert_equal(gottx['txid'], txid3)
+            gottx = self.nodes[n].getrawtransaction(txid=tx, verbose=True, blockhash=block1)
+            assert_equal(gottx['txid'], tx)
             assert_equal(gottx['in_active_chain'], True)
             # We should not get the tx if we provide an unrelated block
-            assert_raises_rpc_error(-5, "No such transaction found", self.nodes[n].getrawtransaction, txid=txid3, blockhash=block2)
+            assert_raises_rpc_error(-5, "No such transaction found", self.nodes[n].getrawtransaction, txid=tx, blockhash=block2)
             # An invalid block hash should raise the correct errors
-            assert_raises_rpc_error(-1, "JSON value is not a string as expected", self.nodes[n].getrawtransaction, txid=txid3, blockhash=True)
-            assert_raises_rpc_error(-8, "parameter 3 must be of length 64 (not 6, for 'foobar')", self.nodes[n].getrawtransaction, txid=txid3, blockhash="foobar")
-            assert_raises_rpc_error(-8, "parameter 3 must be of length 64 (not 8, for 'abcd1234')", self.nodes[n].getrawtransaction, txid=txid3, blockhash="abcd1234")
+            assert_raises_rpc_error(-1, "JSON value is not a string as expected", self.nodes[n].getrawtransaction, txid=tx, blockhash=True)
+            assert_raises_rpc_error(-8, "parameter 3 must be of length 64 (not 6, for 'foobar')", self.nodes[n].getrawtransaction, txid=tx, blockhash="foobar")
+            assert_raises_rpc_error(-8, "parameter 3 must be of length 64 (not 8, for 'abcd1234')", self.nodes[n].getrawtransaction, txid=tx, blockhash="abcd1234")
             foo = "ZZZ0000000000000000000000000000000000000000000000000000000000000"
-            assert_raises_rpc_error(-8, f"parameter 3 must be hexadecimal string (not '{foo}')", self.nodes[n].getrawtransaction, txid=txid3, blockhash=foo)
+            assert_raises_rpc_error(-8, f"parameter 3 must be hexadecimal string (not '{foo}')", self.nodes[n].getrawtransaction, txid=tx, blockhash=foo)
             bar = "0000000000000000000000000000000000000000000000000000000000000000"
-            assert_raises_rpc_error(-5, "Block hash not found", self.nodes[n].getrawtransaction, txid=txid3, blockhash=bar)
+            assert_raises_rpc_error(-5, "Block hash not found", self.nodes[n].getrawtransaction, txid=tx, blockhash=bar)
             # Undo the blocks and verify that "in_active_chain" is false.
             self.nodes[n].invalidateblock(block1)
-            gottx = self.nodes[n].getrawtransaction(txid=txid3, verbose=True, blockhash=block1)
+            gottx = self.nodes[n].getrawtransaction(txid=tx, verbose=True, blockhash=block1)
             assert_equal(gottx['in_active_chain'], False)
             self.nodes[n].reconsiderblock(block1)
             assert_equal(self.nodes[n].getbestblockhash(), block2)
```
</p></details>

jonatack commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -69,24 +74,129 @@ def setup_network(self):
+
+        # Make a tx by sending, then generate 2 blocks; block1 has the tx in it
+        tx = self.nodes[2].sendtoaddress(self.nodes[1].getnewaddress(), 1)
+        block1, block2 = self.nodes[2].generate(2)
+        self.sync_peers()
```
Done

jonatack replied:
Moving the in-mempool transaction tests commit before the simplification commit revealed that removing the `self.sync_peers()` would be flakey without the presence of the in-mempool transaction. This could be fragile, so no longer removing it in the simplification commit.

jonatack commented:
@rajarshimaitra I've appended a commit with you as the author.  Let me know if the name and email `codeShark149 <rajarshi149@gmail.com>` are correct.

fanquake commented:
It could be a good to reorder some of these commits so we're not making one change, then changing the same lines again straight after. For example, in https://github.com/bitcoin/bitcoin/pull/22437/commits/10a3db049ce2858b4885472314567d05d6cc75cb you rename variables i.e `tx`, `txId` to `txid2`, `txid3` etc. However in the following commit (https://github.com/bitcoin/bitcoin/pull/22437/commits/c2d79955a833aec09ad2f7a064509bcdcdbd1cb0), a bunch of those end up being renamed again, i.e `txid3` and `txid2` back to `tx`.

jonatack commented:
@fanquake yes, that's better. Reordered the last three commits and reduced the diff slightly in others.

rajarshimaitra commented:
tACK https://github.com/bitcoin/bitcoin/pull/22437/commits/f0aacf0a87a67f88018f683092f3b7d16e03e6f1

> @rajarshimaitra I've appended a commit with you as the author. Let me know if the name and email codeShark149 <rajarshi149@gmail.com> are correct.

That seems correct. That's very generous of you. You did the work, i just pointed fingers. 

>At first glance I don't see a straightforward way to test which code path is taken in that case with the current code, as there is no observable difference in behavior other than hopefully in performance, for which a benchmark could be added.

Yes that occurred to me too. There is no observational difference between the two paths. So we won't know which one is taken. Probably something for a future improvement. Maybe all it needs is one debug log in the core logic, and with that we can check in the test.

josibake commented:
ACK https://github.com/bitcoin/bitcoin/pull/22437/commits/f0aacf0a87a67f88018f683092f3b7d16e03e6f1

code review (tons of great suggestions from others, so nothing for me to add), compiled and ran the tests locally, switched the `-txindex` nodes to non-txindex and vice-versa to verify tests failed as expected.

overall, major kudos on the refactor. grouping the tests into functions made this much more readable. also +1 for more test coverage

Zero-1729 commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -199,283 +306,25 @@ def run_test(self):
-        self.log.info('sendrawtransaction/testmempoolaccept with maxfeerate')
+    def sendrawtransaction_testmempoolaccept_tests(self):
+        self.log.info("Test sendrawtransaction/testmempoolaccept with maxfeerate")
+        fee_exceeds_max = "Fee exceeds maximum configured by user (e.g. -maxtxfee, maxfeerate)"
```
What's the reason for having this as a variable instead of passing the literal to `assert_raises_rpc_error` as an arg?

jonatack replied:
This variable is invoked twice below in this test, so it clarifies that we check for the same message and also allows reducing the line length to under 120; previously for instance, both lines did not fit in the GitHub display without scrolling (and scrolling is an annoyance with a file or diff of this length, as you have to navigate to the bottom, scroll, then navigate back to the lines).

Zero-1729 replied:
Thanks for clarifying!

Zero-1729 reviewed (approved):
crACK f0aacf0

Nice changes! Local tests also passed 🧉

jonatack commented:
Thanks everyone for the reviews! It would be nice for this to be merged before any other change touching this file invalidates all the review. One can hope :)

jonatack commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -74,12 +75,11 @@ def setup_network(self):
 
     def sync_peers(self):
-        self.sync_mempools()
+        self.sync_mempools(self.nodes[0:3])  # nodes 0 to 3 have a mempool
```
The -blocksonly nodes allow us to add missing test coverage for rpc getrawtransaction error cases like in my https://github.com/bitcoin/bitcoin/pull/22437#discussion_r696812484 above. We could maybe pass non-existing txids instead, but it seems better to test actual -blocksonly behavior.

I have a branch to add that coverage and some other improvements, and rather than increase the size of this pull and invalidate the existing review, I can propose it as a follow-up.

MarcoFalke replied:
> it seems better to test actual -blocksonly behavior.

blocksonly is a network option, so it seems confusing to test it's behaviour inside a raw transaction test.

jonatack replied:
My thought was that -blocksonly effectively disables mempool functionality, which for e2p testing relates to how getrawtransaction behaves when a node operator running -blocksonly calls getrawtransaction, which is valid with either txindex and/or passing blockhash if the tx is in a block, so it seems good to have explicit coverage that serves as a regression test and sanity check. Passing a wrong txid could be done where relevant to verify the same outcome.

MarcoFalke replied:
> -blocksonly effectively disables mempool functionality

The behaviour of the mempool module isn't affected by blocksonly at all. (All txs that are submitted to the mempool are added/rejected in the same way regardless of blocksonly.) blocksonly is purely a net-processing option to disable *incoming* tx relay from network peers. It doesn't affect (in)validity of txs submitted to the mempool. For example, when the wallet or rpc submits a tx, it will still be added to the mempool.

As part of this test, if the goal is to get a fully functioning, but empty mempool, it might be easier to run the tests that you want to run at the beginning with an empty mempool. (Or by restarting a node with `-nopersistmempool` to empty the mempool). Or as mentioned previously by picking a txid that is not currently in the mempool.


If there was an option to completely disable the mempool (i.e. never initialize the class), then it indeed would make a lot of sense to check for that code path.

In testing it is important to be able to draw units/modules and assume that the modules don't interact with each other, other than over their defined interface. This allows to test units/modules separate from each other. If they had to be tested in combination it would lead to path explosion in any real software project, thus making it harder to test meaningful.

Moreover, if the behaviour of one module is required for testing another module, it will make it harder to change the behaviour of that module and only touching the tests of that module.

Or put another way: If `-blockonly` was relevant to test here, why is `-minrelaytxfee` (or any other setting) not relevant to test here?

jonatack replied:
I agree from a unit test point of view, but for end-to-end testing these interactions can be good to test. For example, I think I've found a parameter interaction that doesn't work as documented (not sure yet, need to look further). That seems like good coverage to have if someone proposes it, it is reviewed, and ready to use.  The asmap/addrman interaction is another example.  What to do here?  Lots of time invested so far.  Sure, I could spend even more time to rework this without the coverage and hope for everyone to review it again.  Or just drop the work starting with "Test...with -blocksonly" and maybe pick up the rest later in a follow-up.

MarcoFalke replied:
> The asmap/addrman interaction is another example

I think asmap is part of the addrman module, so testing those together in a test makes sense.

jaysonmald35 replied:
Thank u

jonatack commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -72,17 +73,22 @@ def setup_network(self):
+    def generate_and_sync(self, *, node, blocks, pre_sync=True):
+        if pre_sync == True:
+            self.sync_peers()
+        self.nodes[node].generate(blocks)
```
> [75d1e4c](https://github.com/bitcoin/bitcoin/commit/75d1e4cb14a4bc293a17e516598a1b6b81dcac4a): If you rebase, you can use `self.generate(nodes[node], blocks)`

Thanks. Dropped the commit.

jonatack commented:
Dropped all the commits after d426cab86c1b, no other change not to invalidate review.  Can continue with the other changes in a follow-up.

jonatack commented:
Dropped the commits after d426cab86c1b, no other change to not invalidate review.  Can continue with the other improvements (test de-duplication, in-mempool txn tests, setup simplification, variable naming cleanup, and others) in a follow-up.

MarcoFalke commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -226,28 +227,28 @@ def run_test(self):
-        gottx = self.nodes[0].getrawtransaction(tx, True, block1)
-        assert_equal(gottx['txid'], tx)
-        assert_equal(gottx['in_active_chain'], True)
-        # We should not have the 'in_active_chain' flag when we don't provide a block
```
Not addressed in the latest force push? The first two commits aren't squashed?

jonatack replied:
Thanks for having a look. Didn't change any commits, only dropped the last ones to ~preserve the previous review ACKs.

MarcoFalke replied:
reviewing that two commits are squashed should be as easy as checking that other commits are dropped, no?

jaysonmald35 commented:
Thank u

jonatack commented:
Squashed the first two commits per review feedback and made some minor improvements per `git diff d426cab adfade6`.

mjdietzx commented:
reACK 387355bb9482a09c1fc9b137bea56745a93b7dfd

josibake commented:
reACK https://github.com/bitcoin/bitcoin/pull/22437/commits/387355bb9482a09c1fc9b137bea56745a93b7dfd

maflcko reviewed (commented):
nice. Thanks.



Approach ACK 387355bb9482a09c1fc9b137bea56745a93b7dfd 🔆

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Approach ACK 387355bb9482a09c1fc9b137bea56745a93b7dfd 🔆
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjpnQv/d+jQxCI/GaFHTfHIOmYvfjLAIVoaPnCfJYPoRavcXA92Ov2DiktzzN2j
j8s0Y7Jy86T33aBsbFnGBxM5hWG/WpaceaR7Ga3ixRMQce8TRB9Hng+ab9SuhL7I
eQUdaCAYCDr8pRdY3lmcOOJGzSPNKdiuf2cfFwyDL287VCXrn4JNCTOHZBfjknBu
8aTzGoJXkbbCtxQ26+aNFn4otQrQES9x8VkfrCUbGrUG9wrq64lbhqeoKq3rg6cd
ZbcxL6B8RLVBGiGGVyK4pn0P/sQmoCuND+1fEpUOT7TR7EFmI54Ah3xg3ycsvwx6
WZv+iDh+Zgv/TqTmMvklCHzdxanzxSbnjZ8hu8xiOs3ZVC6r98yYW0Z0QBSx+PN6
WizL0l1tCna5qnRl3ySta74GTOrtFUSihQq9jl+3AWV3Mo3c3Twuz8BCrCXoY+/J
xZEtkIwscpRNKC9nlGzKwItI59+s+B7E3AgmF9gDmOddQ6ZDSwrwrVOrMwuhO9aw
lSJFPNLT
=NEYX
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `2d816e7fc4d23afbf4d13cf22ed4e17e67a0d483e53eeb2099251d3af22c67e7  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401082d816e7fc4d23afbf4d13cf22ed4e17e67a0d483e53eeb2099251d3af22c67e7f0109f33f4a34087a2a78b205d4a7ce573b708fff010312d05cacb401bec8524ab0b2d7391bb08f020864290da483ce023eac8bebd283e3b436b66bff8bab2ae3ff249774ba35a38b008f104612fa72af008753cae07e6040c640083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01040da13048db326c24d4cfa4005db415408f02037dfc5785b5e418605b384d7f2fc3f501df1010de915ce6b4e28cb2ba70588a608f104612fa72af008d06a61ba06f181a50083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010e6ec3b3c2e41ad22cf739c97fb69256308f120794635c6b2b5ba075c28f3f25f692c018ea23a092db1ef3817a9b2bd995c067608f104612fa72af0089d907541340778a90083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -70,23 +73,112 @@ def setup_network(self):
+        vout = find_vout_for_address(self.nodes[1], txid, addr)
+        rawTx = self.nodes[1].createrawtransaction([{'txid': txid, 'vout': vout}], {self.nodes[1].getnewaddress(): 9.999})
+        rawTxSigned = self.nodes[1].signrawtransactionwithwallet(rawTx)
+        txId = self.nodes[1].sendrawtransaction(rawTxSigned['hex'])
```
```
                                   Traceback (most recent call last):
                                     File "/home/marco/workspace/btc_bitcoin_core/test/functional/test_framework/test_framework.py", line 132, in main
                                       self.run_test()
                                     File "/home/marco/workspace/btc_bitcoin_core/test/functional/rpc_rawtransaction.py", line 84, in run_test
                                       self.getrawtransaction_tests()
                                     File "/home/marco/workspace/btc_bitcoin_core/test/functional/rpc_rawtransaction.py", line 107, in getrawtransaction_tests
                                       assert_equal(self.nodes[n].getrawtransaction(txId), rawTxSigned['hex'])
                                     File "/home/marco/workspace/btc_bitcoin_core/test/functional/test_framework/coverage.py", line 49, in __call__
                                       return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
                                     File "/home/marco/workspace/btc_bitcoin_core/test/functional/test_framework/authproxy.py", line 144, in __call__
                                       raise JSONRPCException(response['error'], status)
                                   test_framework.authproxy.JSONRPCException: No such mempool transaction. Use -txindex or provide a block hash to enable blockchain transaction queries. Use gettransaction for wallet transactions. (-5)
```

MarcoFalke replied:
Looks like the test isn't testing anything, as the tx is never included in a block.

If the tx is included, the test will fail:

```diff
diff --git a/test/functional/rpc_rawtransaction.py b/test/functional/rpc_rawtransaction.py
index 96691b2686..b648012413 100755
--- a/test/functional/rpc_rawtransaction.py
+++ b/test/functional/rpc_rawtransaction.py
@@ -99,7 +99,7 @@ class RawTransactionsTest(BitcoinTestFramework):
         rawTx = self.nodes[1].createrawtransaction([{'txid': txid, 'vout': vout}], {self.nodes[1].getnewaddress(): 9.999})
         rawTxSigned = self.nodes[1].signrawtransactionwithwallet(rawTx)
         txId = self.nodes[1].sendrawtransaction(rawTxSigned['hex'])
-        self.generate(self.nodes[0], 1)
+        self.generateblock(self.nodes[0], output=self.nodes[0].getnewaddress(), transactions=[rawTxSigned['hex']])
 
         for n in [0, 3]:
             self.log.info(f"Test getrawtransaction {'with' if n == 0 else 'without'} -txindex")

jonatack replied:
Thanks, was working on the part 2 follow-up to this, will have a look.

jonatack replied:
#24113
